### DISCUSSION TITLE
Combine two arrays on different devices

### DISCUSSION THREAD

**Discussion by mehdiataei** [jax-ml/jax#14145] [General]
I want to concatenate arrays 'a' and 'b' located on devices 0 and 1, respectively, along a specific axis without transferring the data back to a single device (as it would significantly hinder the performance). I am using the new jax.Array and am unsure if device_put_sharded is a viable solution for this task.

Any thoughts?

Thanks

**Comment by mehdiataei** [comment]
Another possible solution would be so somehow only access addressable_data in a sharded array and modify it in-place. But I don't think this is possible in JAX currently. You also cannot access addressable_data in a JITed function.

**Comment by jakevdp** [comment]
If you want to concatenate along the leading axis, effectively creating a sharded array, I suspect the best approach would be using `jax.device_put_sharded`; something like this:
```python
import jax
import numpy as np

devices = jax.devices()
x = jax.device_put(np.zeros(10), device=devices[0])
y = jax.device_put(np.ones(10), device=devices[1])

xy = jax.device_put_sharded([x, y], devices=devices[:2])

print(xy)
# [[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
#  [1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]

print(xy.devices())
# [CpuDevice(id=1), CpuDevice(id=0)]
```

**Reply by mehdiataei** [reply]
Thanks for your answer. 

I want to avoid sharded arrays and utilize automatic parallelization offered by jax.array (however in my use case I need to access parts of it using addressable_data and do some modifications and later on cocat them). 

Also, device_put_sharded acts unexpected a bit. In your example, it is swapping the devices for the arrays thus doing unnecessary communications. I have noticed this too. 

I can modify my question like this:

If we have a jax.array with positional sharding, how do I modify the data of it only one one GPU and keep the same (or a copy) of the positionally sharded jax.array?

**Reply by jakevdp** [reply]
@yashk2810 may be able to answer that

**Reply by yashk2810** [reply]
Sadly you can't do that :) The buffers that an Array consists are immutable.

Why not just use `jnp.concat`? It will do what you want to do via automatic parallelization.

> it is swapping the devices for the arrays

It is actually not. The order of `.devices()` is random because underneath it is a set. To see if you incur any communication, you should look at the HLO rather than the devices.

**Reply by mehdiataei** [reply]
jnp.concat unfortunately doesn't work on arrays that live on different devices. You'll get an error. Am I missing something? My arrays live of two different devices, and all I want is to concat them in a way that I have positional sharding instead of pmap sharding, without any device to device communications.

I'm exploring if I can do it using global device array module. If you have any suggestions I appreciate it.

**Reply by yashk2810** [reply]
You cannot run operations on 2 committed Arrays which are on different devices.

You can leave one of them uncommitted and then the following will work:

```
In [10]: x = jax.device_put(np.zeros(10))
    ...: y = jax.device_put(np.ones(10), device=jax.devices()[1])

In [11]: jnp.concatenate([x, y])
Out[11]:
Array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1., 1.,
       1., 1., 1.], dtype=float32)
```

**Reply by yashk2810** [reply]
> I'm exploring if I can do it using global device array module.

Also jax.Array is the evolution of GlobalDeviceArray. Moreover, GDA is deprecated :)

**Reply by mehdiataei** [reply]
My goal is to avoid the communication between GPUs. The example you provided there is unfortunately communication as the resulting array rests only device 1 without any sharding.

**Reply by yashk2810** [reply]
I think a better thing here would be for you to provide a reproducer and then I can tell you how to modify it.

If 2 Arrays are on the same devices, then jnp.concatenate will still work. But please provide code for what you want to do for me to tell you how to modify things :)

**Reply by mehdiataei** [reply]
Here you go:

```
import os

import jax.numpy as jnp
import jax

# Emulating a multi-device setup
os.environ["XLA_FLAGS"] = '--xla_force_host_platform_device_count=2'

#print(jax.devices())
# Output: [CpuDevice(id=0), CpuDevice(id=1)]

x = jax.device_put(jnp.zeros(8))
y = jax.device_put(jnp.ones(8), device=jax.devices()[1])

r = jnp.concatenate([x, y])
print(r.devices())
# Output: [CpuDevice(id=1)]
# This is not desired, because the array is moved from device 0 to device 1 and has no sharding

'''
What I desire: the output with shape (16,) and sharding on both devices as (2,)
A command such as jax.device_put(jnp.concatenate([x, y]), PositionalSharding(mesh_utils.create_device_mesh((8,)) WITHOUT moving the data first to device 1 and then applying the sharding is desired
Currently, this command does not work, because the array is moved to device 1 first and then the sharding is applied
'''

# For example for r (which already resides on device 1) this works:
from jax.experimental import mesh_utils
from jax.sharding import PositionalSharding

print(r.shape)
r_new = jax.device_put(r, PositionalSharding(mesh_utils.create_device_mesh((2,))))
print(r_new.devices())
# Output: [CpuDevice(id=0), CpuDevice(id=1)]


# But this does not work:

r_1 = r_new.addressable_data(0)
r_2 = r_new.addressable_data(1)

#r_desired = jnp.concatenate([r_1, r_2])
# Output: ValueError: ValueError: Devices of all `Array` inputs and outputs should be the same. Got array device ids [0] on platform CPU and another array's device ids [1] on platform CPU
# We know that concat doesn't work because at least one of tha arrays must be uncommitted. BUT the following also doesn't work:

r_desired = jax.device_put([r_1, r_2], PositionalSharding(mesh_utils.create_device_mesh((2,))))
print(r_desired)

#[Array([0., 0., 0., 0., 0., 0., 0., 0.], dtype=float32), Array([1., 1., 1., 1., 1., 1., 1., 1.], dtype=float32)]
# Each array is on a different device, but the sharding is not applied and we get back a list of arrays

# As I mentioned, usig jax.device_put_sharded, we can get a pmap sharded array back, but that is not desired and I like to get jax.array with positional sharding
```

**Reply by yashk2810** [reply]
I am still unclear on why you don't want the transfer to device 1 happen because it is not going to be expensive at all.

But anyways, you can do this:

```
In [21]: mesh = jax.sharding.Mesh(np.array(jax.devices()).reshape(2,), axis_names=['x',])

In [22]: s = jax.sharding.NamedSharding(mesh, P('x'))

In [23]: @pjit
    ...: def f():
    ...:     x = jnp.zeros(8)
    ...:     y = jax.device_put(jnp.ones(8), jax.devices()[1])
    ...:     out = jnp.concatenate([x, y])
    ...:     return jax.lax.with_sharding_constraint(out, s)
    ...:

In [24]: f().devices()
Out[24]: [CpuDevice(id=0), CpuDevice(id=1)]

In [25]: f().sharding
Out[25]: OpShardingSharding({devices=[2]0,1})
```

Here the SPMD partitioner will propagate the sharding backwards and materialize `x` and `y` on the right devices to begin with so as to not incur any device transfers.

But again, this seems like overkill to me when you can just keep 1 array uncommitted since transferring it to device 1 is not going to be that expensive anyways.

**Reply by mehdiataei** [reply]
Thanks for the example, Yash. I had some issues with a similar approach today, but I'll investigate further tomorrow and update you on the results. 

As for the overkill, when the Array size is +40GB on each device, transferring is either infeasible, or extremely slow 😮‍💨

**Reply by yashk2810** [reply]
> As for the overkill, when the Array size is +40GB on each device, transferring is either infeasible, or extremely slow 😮‍💨

Ahh, I see. This was the missing part :)

**Reply by mehdiataei** [reply]
Hey @yashk2810  and @jakevdp:

This is the solution in case you're wondering:

```
jax.make_array_from_single_device_arrays(shape, sharding, list_of_arrays_on_single_devices)
```
This is really useful in constructing jax Arrays when the array sizes are very large on each GPU.

**Reply by yashk2810** [reply]
Ohh cool! If that works for you, then awesome!

I guess I still didn't understand your question or else i would have recommended that :)
